File: Data Structures & Algorithms/minimum-stack/submission-0.java
class MinStack {

    Stack<Integer> stack = new Stack<Integer>();
    Stack<Integer> minStack = new Stack<Integer>();

    public MinStack() {
        
    }
    
    public void push(int val) {
        stack.push(val);
        if (minStack.empty() || minStack.peek() > val) {
            minStack.push(val);
        } else {
            minStack.push(minStack.peek());
        }
    }
    
    public void pop() {
        stack.pop();
        minStack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
}

File: Data Structures & Algorithms/daily-temperatures/submission-0.kt
class Solution {

    data class Value(val index: Int, val value: Int)

    val stack = Stack<Value>()

    fun dailyTemperatures(temperatures: IntArray): IntArray {
        val result = IntArray(temperatures.size) { 0 }

        for ((index, temperature) in temperatures.withIndex()) {
            while (stack.size > 0 && temperature > stack.peek().value) {
                val top = stack.pop()
                val difference = index - top.index

                result[top.index] = difference
            }

            stack.push(Value(index, temperature))
        }

        return result
    }
}
